examples/simple_enum.rs

```diff
@@ -0,0 +1,1431 @@
+            .truncate(true)
+            .open(&path)
+            .unwrap();
+        bincode::serialize_into(file, inode).unwrap();
+    }
+
+    // Check whether a file should be removed from storage. Should be called after decrementing
+    // the link count, or closing a file handle
+    fn gc_inode(&self, inode: &InodeAttributes) -> bool {
+        if inode.hardlinks == 0 && inode.open_file_handles == 0 {
+            let inode_path = Path::new(&self.data_dir)
+                .join("inodes")
+                .join(inode.inode.0.to_string());
+            fs::remove_file(inode_path).unwrap();
+            let content_path = Path::new(&self.data_dir)
+                .join("contents")
+                .join(inode.inode.0.to_string());
+            fs::remove_file(content_path).unwrap();
+
+            return true;
+        }
+
+        return false;
+    }
+
+    fn truncate(
+        &self,
+        inode: INodeNo,
+        new_length: u64,
+        uid: u32,
+        gid: u32,
+    ) -> Result<InodeAttributes, Errno> {
+        if new_length > MAX_FILE_SIZE {
+            return Err(Errno::EFBIG);
+        }
+
+        let mut attrs = self.get_inode(inode)?;
+
+        check_access(attrs.uid, attrs.gid, attrs.mode, uid, gid, libc::W_OK)?;
+
+        let path = self.content_path(inode);
+        let file = OpenOptions::new().write(true).open(&path).unwrap();
+        file.set_len(new_length).unwrap();
+
+        attrs.size = new_length;
+        attrs.last_metadata_changed = time_now();
+        attrs.last_modified = time_now();
+
+        self.write_inode(&attrs);
+
+        Ok(attrs)
+    }
+
+    fn lookup_name(&self, parent: INodeNo, name: &Path) -> Result<InodeAttributes, Errno> {
+        let entries = self.get_directory_content(parent)?;
+        if let Some((inode, _)) = entries.get(name.as_os_str().as_bytes()) {
+            return self.get_inode(*inode);
+        } else {
+            return Err(Errno::ENOENT);
+        }
+    }
+
+    fn insert_link(
+        &self,
+        req: &Request,
+        parent: INodeNo,
+        name: &Path,
+        inode: INodeNo,
+        kind: FileKind,
+    ) -> Result<(), Errno> {
+        if self.lookup_name(parent, name).is_ok() {
+            return Err(Errno::EEXIST);
+        }
+
+        let mut parent_attrs = self.get_inode(parent)?;
+
+        check_access(
+            parent_attrs.uid,
+            parent_attrs.gid,
+            parent_attrs.mode,
+            req.uid(),
+            req.gid(),
+            libc::W_OK,
+        )?;
+        parent_attrs.last_modified = time_now();
+        parent_attrs.last_metadata_changed = time_now();
+        self.write_inode(&parent_attrs);
+
+        let mut entries = self.get_directory_content(parent).unwrap();
+        entries.insert(name.as_bytes().to_vec(), (inode, kind));
+        self.write_directory_content(parent, entries);
+
+        Ok(())
+    }
+
+    fn dispatch(&mut self, req: &AnyRequest) -> Result<Response, Errno> {
```

Comment:
I like that this returns `Result<Response, Error>` now. Being able to return `Errno` and use Rust's standard handling of errors is way better. And returning `Response` so that it's checked at compile time is a nice improvement too.

I'm less sure about replacing all the `Filesystem` methods with this single `dispatch` method though. There are two main downsides that I see:
1) users are forced to add this really big `match` clause. They can of course delegate to helper methods, but then it's pretty much like the old interface.
2) adding a new enum value with default handling is complicated. The current `Filesystem` trait has default handling for many methods, since there are a bunch that don't really need to be implemented for basic filesystems. I guess we can export a function like `default_handler`, but it seems less nice to make users have a default clause in the `match` which calls it.

For the goal of making ABI changes backwards compatible (which I like, btw!), how about changing `Filesystem` so that all the methods take a single argument which is the the enum struct used here, and they would all return `Result<Response, Errno>`? To preserve backwards compatibility with the existing API, we could introduce it as `Filesystem2` to start with, and deprecate `Filesystem`.

Reply:
> 1. users are forced to add this really big `match` clause. They can of course delegate to helper methods, but then it's pretty much like the old interface.

The difference with delegation with the enum API is that you can delegate multiple operations at once, rather than having to delegate each operation separately.

For example: I have a filesystem where I want to handle `/` differently to `by-commit/*`, also differently to `by-ref/**`.  With the enum I can essentially do:

```rust
fn handler(req: AnyRequest) -> Result<Response, Errno> {
    let inode_type = req.inode_no() & 0xf000_0000_0000_0000 >> 60;
    match inode_type {
        0 => handler_root(req),
        1 => handler_by_commit(req),
        2 => handler_by_ref(req),
        ...
        _ => Err(Errno::ENOENT),
    }
}
```

Similarly if I have some helper (like a middleware) that wants to handle some requests, but not all:

```rust
fn my_helper(req: AnyRequest) -> Result<Response, Errno>;

fn handler(req: AnyRequest) -> Result<Response, Errno> {
    match {
        op::Forget(_) | op::Lookup(_) => return my_helper(),
        _ => (),
    }

    ....
}
```

I don't need to add that code to each of my trait method implementations.

> 2. adding a new enum value with default handling is complicated. The current `Filesystem` trait has default handling for many methods, since there are a bunch that don't really need to be implemented for basic filesystems. I guess we can export a function like `default_handler`, but it seems less nice to make users have a default clause in the `match` which calls it.

I hadn't considered this. My assumption was that all operations return `ENOSYS` by default, so I was thinking that you'd just have a `_ => Err(Errno::ENOSYS)` handler at the bottom of the `match`, but I like your idea of delegating to a `default_handler` instead. The nice thing is that the user can delegate all the operations they choose not to handle in a single line.

> For the goal of making ABI changes backwards compatible (which I like, btw!), how about changing `Filesystem` so that all the methods take a single argument which is the the enum struct used here, and they would all return `Result<Response, Errno>`? To preserve backwards compatibility with the existing API, we could introduce it as `Filesystem2` to start with, and deprecate `Filesystem`.

If we wanted a async or a multithreaded version we'd also need `FilesystemAsync` where each method returns `Future<...>` and `FilesystemMultiThreaded` where each method takes `&self` rather than `&mut self`.  The enum API removes the coupling of how the user code is called from what operations are available, so instead there is no duplication.  It also has advantages WRT async methods in traits.  We don't need to box and use &dyn for every trait method.

Reply:
> If we wanted a async or a multithreaded version we'd also need FilesystemAsync where each method returns Future<...> and FilesystemMultiThreaded where each method takes &self rather than &mut self. The enum API removes the coupling of how the user code is called from what operations are available, so instead there is no duplication. It also has advantages WRT async methods in traits. We don't need to box and use &dyn for every trait method.

This is definitely a strong argument. It'd be great to have async without duplicating a bunch of code! Are you sure this will work though? These enums all contain borrows, and so I feel like that won't play nicely with async, in that all the lifetimes will need to be `'static`.

p.s.
I'm traveling until next week, so may be a little slow to reply

Reply:
> Are you sure this will work though? These enums all contain borrows, and so I feel like that won't play nicely with async, in that all the lifetimes will need to be 'static.

I'm pretty sure. The data will still be owned by the future higher up in the async call stack.  The code would look something like:

```rust
pub async fn serve_async<Handler>(chan: Channel, handler: Handler) -> io::Result<()>
where
    Handler: Fn(AnyRequest<'_>) -> Result<Response, Errno>,
{
    loop {
        let mut buf = vec![];
        chan.read_into(buf).await?;
        spawn(handle_one(buf, handler.clone()))
    }
}

async fn handle_one<Handler>(buf: Vec<u8>, handler: Handler) -> Result<Response, Errno>
where
    Handler: Fn(AnyRequest<'_>) -> Result<Response, Errno>,
{
    let req = AnyRequest::try_from(buf)?;
    handler(req).await
}
```

So the buffer is owned by an async stack frame higher up in the async stack. References to that buffer can be passed to lower stack frames with a non-static lifetime.  Of course this is not a stack in the traditional sense. rustc will transform the `handle_one` async fn into a self-referential struct that looks like:

```rust
struct HandleOneFuture {
    buf: Vec<u8>,
    req: AnyRequest<'buf>,
} 
```

but we don't have to worry about such details.

I don't see any reason that this shouldn't work, but I don't have code that actually does this yet.  Baby-steps and all that.

Reply:
Ah ok, I think you're right. My grasp of the borrow checker is still a little shaky :) I played around in Rust Playground, and the thing that I thought wouldn't work seems to work fine, so ya let's do!